sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/SupportsNamespaces.java

```diff
@@ -0,0 +1,145 @@
+ */
+public interface SupportsNamespaces extends CatalogPlugin {
+
+  /**
+   * Return a default namespace for the catalog.
+   * <p>
+   * When this catalog is set as the current catalog, the namespace returned by this method will be
+   * set as the current namespace.
+   * <p>
+   * The namespace returned by this method is not required to exist.
+   *
+   * @return a multi-part namespace
+   */
+  default String[] defaultNamespace() {
+    return new String[0];
+  }
+
+  /**
+   * List top-level namespaces from the catalog.
+   * <p>
+   * If an object such as a table, view, or function exists, its parent namespaces must also exist
+   * and must be returned by this discovery method. For example, if table a.b.t exists, this method
+   * must return ["a"] in the result array.
+   *
+   * @return an array of multi-part namespace names
+   */
+  String[][] listNamespaces() throws NoSuchNamespaceException;
+
+  /**
+   * List namespaces in a namespace.
+   * <p>
+   * If an object such as a table, view, or function exists, its parent namespaces must also exist
+   * and must be returned by this discovery method. For example, if table a.b.t exists, this method
+   * invoked as listNamespaces(["a"]) must return ["a", "b"] in the result array.
+   *
+   * @param namespace a multi-part namespace
+   * @return an array of multi-part namespace names
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   */
+  String[][] listNamespaces(String[] namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Test whether a namespace exists.
+   * <p>
+   * If an object such as a table, view, or function exists, its parent namespaces must also exist.
+   * For example, if table a.b.t exists, this method invoked as namespaceExists(["a"]) or
+   * namespaceExists(["a", "b"]) must return true.
+   *
+   * @param namespace a multi-part namespace
+   * @return true if the namespace exists, false otherwise
+   */
+  default boolean namespaceExists(String[] namespace) {
+    try {
+      loadNamespaceMetadata(namespace);
+      return true;
+    } catch (NoSuchNamespaceException e) {
+      return false;
+    }
+  }
+
+  /**
+   * Load metadata properties for a namespace.
+   *
+   * @param namespace a multi-part namespace
+   * @return a string map of properties for the given namespace
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   * @throws UnsupportedOperationException If namespace properties are not supported
+   */
+  Map<String, String> loadNamespaceMetadata(String[] namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Create a namespace in the catalog.
+   *
+   * @param namespace a multi-part namespace
+   * @param metadata a string map of properties for the given namespace
+   * @throws NamespaceAlreadyExistsException If the namespace already exists
+   * @throws UnsupportedOperationException If create is not a supported operation
+   */
+  void createNamespaceMetadata(
```

Comment:
`createNamespace` instead of `createNamespaceMetadata`

Reply:
I recently changed this and I think it fits better with the semantics that we require. Namespaces are required to exist if they contain a table. We don't check whether the namespace exists on table creation and leave that to the implementation. So it makes more sense if this is create namespace metadata, because the namespace may already exist and we don't want to throw NamespaceAlreadyExists if there isn't already metadata. Does that make sense?

Reply:
See this test: https://github.com/apache/spark/pull/24560/files#diff-dbc34e865cad1087fa8c24afe5e5e1aeR761-R773

Because the table exists, the namespace already exists. It would be surprising to fail the create because it was implicitly created earlier, and it would also be surprising to allow a create when it already exists. Adding "metadata" to the method name makes this more obvious because you can create the namespace metadata.

Reply:
got it, that's fair.

Reply:
Do you mean sometimes `createNamespaceMetadata` is the same as adding properties to an existing namespace?

Reply:
@cloud-fan, look at the test case that I linked to. In that case, the namespace exists because it contains a table. Calling `createNamespaceMetadata` is allowed because it doesn't have metadata associated with it, which makes it exist independent of tables. So, yes, this does function like adding properties the first time.

Reply:
This behavior is a little counter-intuitive to me. When I run `mkdir -p /dir1/dir2/dir3` to create a bunch of directories, `mkdir /dir1` fails even if `dir1` was created implicitly.

It also complicates the abstraction. Now namespace has 2 different kinds: created explicitly and created implicitly. The one created explicitly will make `createNamespaceMetadata` fail, and the one created implicitly will make `createNamespaceMetadata` success.

Reply:
@cloud-fan, that's why this was renamed to `createNamespaceMetadata`, not `createNamespace`.

Reply:
It seems to me that a simple `createNamespace` method is more useful, instead of a `createNamespaceMetadata` method that needs more document to clarify the behavior.

Reply:
@cloud-fan, if we do that, then do you think implementations that extend `SupportsNamespaces` should fail to create a table if its namespace doesn't exist? I think that's a bad idea -- think about file system tables where directories are namespaces and may not exist.

I understand not liking the method name, but we need some way to solve this problem. Curious about what you'd suggest.

Reply:
If you run `CREATE db.tbl ...` in JDBC, it will complain about `db` not exist. This is the same in Spark, Hive, Preso and RDBMS like postgres.

I think it's more common to fail the table creation if the namespace doesn't exist. Even for file systems, `mkdir` fails if the parent directory doesn't exist, unless you add the `-p` option.

We can add a variant of `createTable` method in the future, which is similar to `mkdir -p` and can create the missing namespaces automatically. By default I think table creation should fail if the namespace doesn't exist, to simplify the semantic.

Reply:
@cloud-fan, some sources like JDBC and Hive will throw an exception when the namespace doesn't exist. The problem is that we don't want to require this for all sources where it doesn't make sense, like a catalog that stores tables in an object store. Object stores don't necessarily have the concept of a directory that exists independently (s3 FS wrappers have conventions like `_$folder$` objects).

That's why failing to create a table if the namespace doesn't exist is optional, and creating namespaces implicitly is allowed. This isn't something we can (or should) change.

So the question is how do we handle this in this API? I think that it makes sense to allow namespace creation that is independent of tables -- by creating namespace metadata.

Reply:
Instead of `createNamespaceMetadata` can we make this something like `tryCreateNamespace`, which is not required to throw an exception if the namespace already exists? Or, `getOrCreateNamespace`, for a similar effect?

Reply:
@rdblue If the namespace already exists (implicitly for that matter), wouldn't you use alterNamespace to add the metadata?

Reply:
@brkyvz, that also works (and has a test in this PR).

@mccheah, I would be surprised if `tryCreateNamespace` and `getOrCreateNamespace` updated the namespace metadata, which I think is part of the intent here.

Another option is `ensureNamespaceExists` without metadata, which would create or do nothing if it already exists. Then metadata could be added using `alterNamespace`. I don't think that is a very good option either because there is no clean create with metadata operation.

Reply:
I'm not a fan of "implementation is free to decide the behaviors". When end-users run SQL queries in Spark, Spark should behave deterministically and consistently, instead of telling end-users "you will see different behaviors if you connect to different catalogs".

IMO, `CREATE TABLE` should fail if the namespace doesn't exist, as it's intuitive and commonly seen in other databases. If a catalog stores tables in an object store, and is hard to detect namespace existence, then it should not support namespace, or try harder to figure out a workaround.

We can add an option/SQL syntax to allow `CREATE TABLE` to create namespaces automatically, but it should be the end-users who decide which behavior to pick, not the catalog. This is similar to `mkdir`. Users must add the `-p` option to create non-existing parent directories. It's pretty bad if we say "mkdir may or may not create parent directories, it depends on which Linux distribution you run it".

Reply:
@cloud-fan But the whole goal of DSV2 is to let the implementation decide the (internal) behavior, but have Spark to enforce the semantics. If you truncate a table with Spark, all Spark needs is that the next time that table is queried that there is no data in the table. It doesn't care if the table (datasource) decided to physically delete all the data or just did a logical deletion.

 - `CREATE TABLE` should fail if the namespace doesn't exist => This makes sense to me. I'd say this is the behavior that users expect. I don't think this would be a blocker if people start to implement object store backed catalogs. Then they can run `CREATE NAMESPACE a.b.c`, and then try the CREATE TABLE again
 - If a namespace already exists, then IMHO CREATE NAMESPACE should throw an error and they should run ALTER NAMESPACE to add the metadata
 - Regarding implicit namespaces in object stores, I feel that this still means that someone created the namespace through some other means (imagine using a GUI to create a MySQL table instead of the shell), and alter namespace should be used to add the metadata.

wdyt @rdblue @mccheah

Reply:
@brkyvz when I say "behavior", I don't mean the internal behavior that is invisible to end-users, I actually mean the semantic that is visible to end-users.